cpp/src/strings/datetime.cu

```diff
@@ -0,0 +1,733 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/wrappers/timestamps.hpp>
+#include <cudf/strings/datetime.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/utilities/error.hpp>
+#include "./utilities.hpp"
+#include "./utilities.cuh"
+
+#include <vector>
+#include <map>
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/sort.h>
+#include <thrust/sequence.h>
+
+namespace cudf
+{
+namespace strings
+{
+namespace
+{
+
+/**
+ * @brief  Units for timestamp conversion.
+ * These are defined since there are more than what cudf supports.
+ */
+enum timestamp_units {
+    years,           ///< precision is years
+    months,          ///< precision is months
+    days,            ///< precision is days
+    hours,           ///< precision is hours
+    minutes,         ///< precision is minutes
+    seconds,         ///< precision is seconds
+    ms,              ///< precision is milliseconds
+    us,              ///< precision is microseconds
+    ns               ///< precision is nanoseconds
+};
+
+
+// used to index values in a timeparts array
+enum timestamp_parse_component {
+    TP_YEAR        = 0,
+    TP_MONTH       = 1,
+    TP_DAY         = 2,
+    TP_HOUR        = 3,
+    TP_MINUTE      = 4,
+    TP_SECOND      = 5,
+    TP_SUBSECOND   = 6,
+    TP_TZ_MINUTES  = 7,
+    TP_ARRAYSIZE   = 8
+};
+
+struct format_item
```

Comment:
This should probably have an `alignas()` if you are going to create device arrays of them.